controller/monitor/snapshot_monitor.go

```diff
@@ -0,0 +1,588 @@
+	}
+
+	return nil
+}
+
+func (m *SnapshotMonitor) checkVolumeNotInMigration(volumeName string) error {
+	v, err := m.ds.GetVolume(volumeName)
+	if err != nil {
+		return err
+	}
+	if v.Spec.MigrationNodeID != "" {
+		return fmt.Errorf("cannot operate during migration")
+	}
+	return nil
+}
+
+func (m *SnapshotMonitor) getEngineBinaryClient(volumeName string) (client *engineapi.EngineBinary, err error) {
+	var e *longhorn.Engine
+
+	defer func() {
+		err = errors.Wrapf(err, "cannot get client for volume %v", volumeName)
+	}()
+	es, err := m.ds.ListVolumeEngines(volumeName)
+	if err != nil {
+		return nil, err
+	}
+	if len(es) == 0 {
+		return nil, fmt.Errorf("cannot find engine for volume %v", volumeName)
+	}
+	if len(es) != 1 {
+		return nil, fmt.Errorf("more than one engine for volume %v exist", volumeName)
+	}
+	for _, e = range es {
+		break
+	}
+	if e.Status.CurrentState != longhorn.InstanceStateRunning {
+		return nil, fmt.Errorf("engine %v is not running", e.Name)
+	}
+	if isReady, err := m.ds.CheckEngineImageReadiness(e.Status.CurrentImage, m.nodeName); !isReady {
+		if err != nil {
+			return nil, fmt.Errorf("cannot get engine client with image %v: %v", e.Status.CurrentImage, err)
+		}
+		return nil, fmt.Errorf("cannot get engine client with image %v because it isn't deployed on this node", e.Status.CurrentImage)
+	}
+
+	engineCollection := &engineapi.EngineCollection{}
+	return engineCollection.NewEngineClient(&engineapi.EngineClientRequest{
+		VolumeName:  e.Spec.VolumeName,
+		EngineImage: e.Status.CurrentImage,
+		IP:          e.Status.IP,
+		Port:        e.Status.Port,
+	})
+}
+
+func (m *SnapshotMonitor) syncHashStatusFromEngineReplicas(engine *longhorn.Engine, engineClientProxy engineapi.EngineClientProxy,
+	snapshotName string) error {
+	hashStatus, err := engineClientProxy.SnapshotHashStatus(engine, snapshotName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get hash status for snapshot %v", snapshotName)
+	}
+
+	for _, status := range hashStatus {
+		if status.State == string(SnapshotHashStatusError) {
+			return fmt.Errorf("failed to hash snapshot %v since %v", snapshotName, status.Error)
+		}
+
+		if status.State == string(engineapi.ProcessStateInProgress) {
+			return fmt.Errorf(string(engineapi.ProcessStateInProgress))
+		}
+	}
+
+	snapshot, err := m.ds.GetSnapshot(snapshotName)
+	if err != nil {
+		return err
+	}
+	existingSnapshot := snapshot.DeepCopy()
+
+	checksum, err := determineChecksumFromHashStatus(snapshotName, snapshot.Status.Checksum, hashStatus)
+	if err != nil {
+		m.eventRecorder.Eventf(engine, v1.EventTypeWarning, constant.EventReasonFailedSnapshotDataIntegrityCheck,
+			"Failed to check the data integrity of snapshot %v for volume %v", snapshotName, engine.Spec.VolumeName)
+		return errors.Wrapf(err, "failed to determine checksum for snapshot %v", snapshotName)
+	}
+
+	snapshot.Status.Checksum = checksum
```

Comment:
Just realized that the hashing state and progress are not recorded in the Snapshot CRs. Do we need to record it?

Reply:
I think we don't need to record the transisent information. We don't plan to expose the snapshot checksum information to users explicitly in the first version.